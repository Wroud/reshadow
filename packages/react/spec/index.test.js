/**
 * @jest-environment enzyme
 */

import {render, shallow} from 'enzyme';

const getStyles = () =>
    [...document.getElementsByTagName('style')]
        .map(x => x.textContent)
        .join('');

/** @jsx jsx */
let styled, css, jsx;

describe('react', () => {
    beforeEach(() => {
        [...document.getElementsByTagName('style')].forEach(x => {
            x.remove();
        });

        jest.isolateModules(() => {
            ({default: styled, css, jsx} = require('..'));
        });
    });

    it('should apply styles', () => {
        const Button = ({children}) => styled`
            button {
                color: red;
            }
        `(<button>{children}</button>);

        const wrapper = render(<Button>click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply dynamic styles', () => {
        const Button = ({children, color}) => styled`
            button {
                color: ${color};
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button color="red">click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({color: 'green'});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply external styles', () => {
        const styles = css`
            button {
                color: red;
            }
        `;

        const Button = ({children}) => styled(styles)(
            <button>{children}</button>,
        );

        const wrapper = render(<Button>click me</Button>);
        expect(wrapper).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply external dynamic styles', () => {
        const styles = ({color}) => css`
            button {
                color: ${color};
            }
        `;

        const Button = ({children, color}) => styled(styles({color}))(
            <button>{children}</button>,
        );

        const wrapper = shallow(<Button color="red">click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({color: 'green'});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply global styles', () => {
        const styles = ({color}) => css`
            :global(button) + :global(button) {
                color: ${color};
            }
        `;

        const Button = ({children, color}) => styled(styles({color}))(
            <button>{children}</button>,
        );

        const wrapper = shallow(<Button color="red">click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply mixins', () => {
        const theme = css`
            color: red;
            border: none;
        `;

        const Button = ({children}) => styled`
            button {
                ${theme}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button>click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply mixins with pseudo', () => {
        const theme = css`
            color: red;
            border: none;
            &:focus {
                color: green;
            }
        `;

        const Button = ({children}) => styled`
            button {
                ${theme}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button>click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply dynamic mixins', () => {
        const theme = ({color}) => css`
            color: ${color};
            border: none;
        `;

        const Button = ({children, color}) => styled`
            button {
                ${theme({color})}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button color="red">click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({color: 'green'});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply conditional mixins', () => {
        const theme = ({variant, color}) =>
            variant === 'inverse'
                ? css`
                      color: ${color};
                      background: white;
                  `
                : css`
                      color: white;
                      background: ${color};
                  `;

        const Button = ({children, variant, color}) => styled`
            button {
                ${theme({color, variant})}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(
            <Button variant="normal" color="red">
                click me
            </Button>,
        );
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();

        wrapper.setProps({color: 'green', variant: 'inverse'});
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply composed mixins', () => {
        const padding = css`
            padding: 5px 10px;
        `;

        const theme = ({color}) => css`
            color: ${color};
            ${padding}
        `;

        const Button = ({children, color}) => styled`
            button {
                ${theme({color})}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button color="red">click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should apply object mixins', () => {
        const padding = {
            padding: '5px 10px',
        };

        const Button = ({children}) => styled`
            button {
                ${padding}
                margin: 10px;
            }
        `(<button>{children}</button>);

        const wrapper = shallow(<Button>click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });

    it('should work with Fragments', () => {
        const Button = ({children}) => styled`
            button {
                margin: 10px;
            }
        `(
            <>
                <button>{children}</button>
            </>,
        );

        const wrapper = shallow(<Button>click me</Button>);
        expect(wrapper.render()).toMatchSnapshot();
        expect(getStyles()).toMatchSnapshot();
    });
});
