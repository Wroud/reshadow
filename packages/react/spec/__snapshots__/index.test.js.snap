// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react should apply composed mixins 1`] = `
<button
  class="__button_lm3xvp"
  style="--1owzm37_1:red"
>
  click me
</button>
`;

exports[`react should apply composed mixins 2`] = `".__button_lm3xvp{color:var(--1owzm37_1);padding:5px 10px;margin:10px;}"`;

exports[`react should apply conditional mixins 1`] = `
<button
  class="__button_nzpf5v"
  style="--14y9117_1:red"
>
  click me
</button>
`;

exports[`react should apply conditional mixins 2`] = `".__button_nzpf5v{color:white;background:var(--14y9117_1);margin:10px;}"`;

exports[`react should apply conditional mixins 3`] = `
<button
  class="__button_1pj7rgr"
  style="--134zod7_1:green"
>
  click me
</button>
`;

exports[`react should apply conditional mixins 4`] = `".__button_nzpf5v{color:white;background:var(--14y9117_1);margin:10px;}.__button_1pj7rgr{color:var(--134zod7_1);background:white;margin:10px;}"`;

exports[`react should apply dynamic mixins 1`] = `
<button
  class="__button_1hzemul"
  style="--xqnn5g_1:red"
>
  click me
</button>
`;

exports[`react should apply dynamic mixins 2`] = `".__button_1hzemul{color:var(--xqnn5g_1);border:none;margin:10px;}"`;

exports[`react should apply dynamic mixins 3`] = `
<button
  class="__button_1hzemul"
  style="--xqnn5g_1:green"
>
  click me
</button>
`;

exports[`react should apply dynamic mixins 4`] = `".__button_1hzemul{color:var(--xqnn5g_1);border:none;margin:10px;}"`;

exports[`react should apply dynamic styles 1`] = `
<button
  class="__button_1pacrc6"
  style="--xkvkx5_1:red"
>
  click me
</button>
`;

exports[`react should apply dynamic styles 2`] = `".__button_1pacrc6{color:var(--xkvkx5_1);}"`;

exports[`react should apply dynamic styles 3`] = `
<button
  class="__button_1pacrc6"
  style="--xkvkx5_1:green"
>
  click me
</button>
`;

exports[`react should apply dynamic styles 4`] = `".__button_1pacrc6{color:var(--xkvkx5_1);}"`;

exports[`react should apply external dynamic styles 1`] = `
<button
  class="__button_1pacrc6"
  style="--xkvkx5_1:red"
>
  click me
</button>
`;

exports[`react should apply external dynamic styles 2`] = `".__button_1pacrc6{color:var(--xkvkx5_1);}"`;

exports[`react should apply external dynamic styles 3`] = `
<button
  class="__button_1pacrc6"
  style="--xkvkx5_1:green"
>
  click me
</button>
`;

exports[`react should apply external dynamic styles 4`] = `".__button_1pacrc6{color:var(--xkvkx5_1);}"`;

exports[`react should apply external styles 1`] = `
<button
  class="__button_olz31h"
>
  click me
</button>
`;

exports[`react should apply external styles 2`] = `".__button_olz31h{color:red;}"`;

exports[`react should apply global styles 1`] = `
<button
  style="--aku94k_1:red"
>
  click me
</button>
`;

exports[`react should apply global styles 2`] = `"button + button{color:var(--aku94k_1);}"`;

exports[`react should apply mixins 1`] = `
<button
  class="__button_13d2oc0"
>
  click me
</button>
`;

exports[`react should apply mixins 2`] = `".__button_13d2oc0{color:red;border:none;margin:10px;}"`;

exports[`react should apply mixins with pseudo 1`] = `
<button
  class="__button_1tuwmo"
>
  click me
</button>
`;

exports[`react should apply mixins with pseudo 2`] = `".__button_1tuwmo:focus{color:green;}.__button_1tuwmo{color:red;border:none;margin:10px;}"`;

exports[`react should apply object mixins 1`] = `
<button
  class="__button_qjpl95"
>
  click me
</button>
`;

exports[`react should apply object mixins 2`] = `".__button_qjpl95{padding:5px 10px;margin:10px;}"`;

exports[`react should apply styles 1`] = `
<button
  class="__button_olz31h"
>
  click me
</button>
`;

exports[`react should apply styles 2`] = `".__button_olz31h{color:red;}"`;
